### ISSUE TITLE
Best way to solve unique error handling using promises

### ISSUE THREAD

**Issue by joaufi** [joaufi/up-text-front-end#1]
I am trying to figure out the best way to solve the following problem:

I have a promise chain of `.then()` and `.catch()` when I am creating or editting tasks in my to-do list. I would like it so that if the FIRST promise (api call) on failure makes it so that no following promises execute and no `.then()` or `.catch()` run. 

I have devised a way to do this like so: 
```js
// tevents.js 

const onCreateTask = function (event) {
  event.preventDefault();
  let data = getFormFields(this);
  data.task.complete = false;
  tapi.createTask(data)
    .then((response_data) => {
      tui.createTaskSuccess(response_data);
      tapi.indexTasks()
        .then(tui.indexTasksSuccess)
        .catch(tui.failure);
    })
    .catch(tui.createTaskFailure);
};
```

However, I was warned away from doing this as it could lead to Promise hell much like Callback hell. It works but I would like to do it in a better fashion.

Another way that was recommended to me was to make a generic `tui.failure` which would then check the type of error it is passed and do different things based on the type of error. This seems like a fine solution but I'm not totally sold on it. I would much rather have unique failure functions for each event, as I feel this is much more extendable and easier to manage as features are added in the future.

I'm basically just looking for some guidance on how I might best go about doing this failure handling. Anyone have any ideas?

**Comment by astoellis** [comment]
I might put the "update tasks" functionality into a different function, but that really just makes it _look_ cleaner, it may not be of much value to you. However, it does satisfy a separation of concerns, and take you up one level of promise hell (visually).

My naive solution (not 100% sure if this'll work) would be something like this:

```js
//tevents.js
function triggerListChange() {
  tapi.indexTasks()
    .then(tui.indexTasksSuccess)
    .catch(tui.failure);
}

const onCreateTask = function (event) {
  event.preventDefault();
  let data = getFormFields(this);
  data.task.complete = false;
  tapi.createTask(data)
    .then(tui.createTaskSuccess )
    .then( triggerListChange )
    .catch(tui.createTaskFailure);
};
```

Again, this is kinda the difficulty with async code, so it's good you're being introduced to it now. Hope this helps some!

**Comment by gaand** [comment]
What are you going to do differently in the different error handlers?

Why do you need to call the api a second time?

**Comment by joaufi** [comment]
@gaand 

```js
const createTaskFailure = (data) => {
  Materialize.toast('There was a problem with the data you entered, take a look and try again!', 4000);
};
```

So that failure function would actually work for both createTask and editTask. However, I would like the generic `ui.failure` to post a different message than the createTask or editTask failure messages, so the user knows better what is failing.

Ideally I'd like to make unique failure states for events and task events that make sense to have them, so that I can provide the user with more specific information about what may be going wrong.

As for your second question, do you mean why do I call `tapi.indexTasks()`? I do that after a successful edit or create to make sure the changes to the backend immediately appear on the front-end. My indexTasksSuccess ui function reloads my main content div with a handlebar template with the updated data.

**Comment by joaufi** [comment]
This is what I am currently thinking of doing as per @astoellis recommendation
```js
const onIndexTasks = function () {
  tapi.indexTasks()
    .then(tui.indexTasksSuccess)
    .catch(tui.failure);
};

const onCreateTask = function (event) {
  event.preventDefault();
  let data = getFormFields(this);
  data.task.complete = false;
  tapi.createTask(data)
    .then(tui.createTaskSuccess)
    .then(onIndexTasks)
    .catch(tui.createTaskFailure);
};
```

I don't really see any issue with this but I have not been working with these for long so I might be missing something.

**Comment by gaand** [comment]
A better phrasing of my second question is why are you asking the api for data you already have in the client?

**Comment by joaufi** [comment]
Oh, okay. So, On create/update/destroy task the task is modified in the backend database. Since the backend database changes, I thought the best way to show these changes to the user on the front end would be to simply re-call indexTasks which pulls all the rows in the tasks table and passes them to a Handlebar template which then displays them, thus updating the page.

I suppose I could store the data that I get from the initial indexTasks and then create/update/destroy the task both in the frontend representation of the data and the backend but then what is the point of having a backend? At least that's how I'm thinking about it. I could very well be wrong.

**Comment by gaand** [comment]
It looks like the code above would produce the correct result.

This probably would as well

```js
  const createTaskPromise = tapi.createTask(data);

  createTaskPromise
    .then(tui.createTaskSuccess)
    .catch(tui.createTaskFailure);

  createTaskPromise
    .then(tapi.indexTasks)
    .then(tui.indexTasksSuccess)
    .catch(tui.failure);

```

That solution would decouple how you handle `tapi.indexTasks` errors from where you're using it.

**Comment by gaand** [comment]
Create is different.  You need the data from the back end.

Update and destroy you don't, since you should already have it.

The point of the back end is to persist and share data across machines, users, whatever.

When Ia user logs into a different machine, should they be able to see what they did from their phone?

**Comment by joaufi** [comment]
Yeah, if a user logs on from another device they should definitely be able to see if they updated or destroyed a task. The changes should persist across all points a user can login to the website.

It just seems to me if I update a task or destroy it, I'm going to have to modify that in the backend anyway if I want the user to be able to see the changes they made on other devices.

That solution you posted is good though. I'm all about decoupling things.

**Comment by gaand** [comment]
The point of not making the second request is that network requests can go wrong for all sorts of reasons. But updating the DOM from data you already have is robust.  

So create returns the new object and you add it to what's already there. Similarly for update and destroy.

Refetching everything is the brute force way.

**Comment by joaufi** [comment]
Got it! That is really good to know, thanks @gaand . I had no idea. I will work on switching things over on a different branch, as it works for resubmission for now.

**Comment by gaand** [comment]
Good plan.  I didn't realize this was your project two.  I might have advised differently until after your resubmit 🤖 😄
